### PR TITLE
Track RetryAttempt and Phase of ExternalResources

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
Populating RetryAttempt and Phase on ExternalResources to track array subtasks.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
Created a new [ExternalResources](https://github.com/flyteorg/flyteplugins/blob/de5b4773187ced71b133e822bccc3fc3c8a85f16/go/tasks/pluginmachinery/core/phase.go#L70) struct which contains metadata on, you may have guessed, external resources which replaces the Metadata protobuf stored in [TaskInfo](https://github.com/flyteorg/flyteplugins/blob/973a51ae8cc6a06e2f78007fe1ffda71f7355fad/go/tasks/pluginmachinery/core/phase.go#L81). This decision was made for two reasons (1) the phase maintained within flyteplugins [for each subtask](https://github.com/flyteorg/flyteplugins/blob/973a51ae8cc6a06e2f78007fe1ffda71f7355fad/go/tasks/plugins/array/k8s/monitor.go#L87) is a flyteplugins core.Phase rather than the TaskExecution.Phase from flyteidl (which is reported in the TaskExecutionEvent) and (2) in flytepropeller all the Metadata fields except external resources [are being overwritten](https://github.com/flyteorg/flytepropeller/blob/51c85b7b842e51489b158fb7dcddaefd04f98dec/pkg/controller/nodes/task/transformer.go#L105), plus we need to convert between the flyteplugin and flyteidl phases.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1838

## Follow-up issue
_NA_

### CODE REVIEW

**Review by EngHabu — DISMISSED** [review]
Looks good to me... Before we merge though. Can you build docker image for propeller with these changes so we can test on demo?

**Review by EngHabu — APPROVED** [review]
